### DISCUSSION TITLE
Dynamic simulation using MFrontNonlinearProblem

### DISCUSSION THREAD

**Discussion by VaishnavBhaskaran** [thelfer/tfel#618] [Q&A] [answered]
Hello @thelfer,

I set up the dynamic simulation of my problem setup from #601 based on [Elastodynamics tutorial](https://fenicsproject.org/olddocs/dolfin/2019.1.0/python/demos/elastodynamics/demo_elastodynamics.py.html) and [Transient heat equation with phase change](https://thelfer.github.io/mgis/web/mgis_fenics_heat_equation_phase_change.html).

I define the constants and setup the `MFrontNonlinearMaterial` and `MFrontNonlinearProblem` as shown below

```python
eta_m = Constant(0.)
eta_k = Constant(0.)

alpha_m = Constant(0.2)
alpha_f = Constant(0.4)
gamma = Constant(0.5+alpha_f-alpha_m)
beta = Constant((gamma+0.5)**2/4.)

T = 1.1
N_steps = 60
dt = Constant(0.)

rho = Constant(8e-6)
mat_prop = {"YoungModulus": 185000,
        "PoissonRatio": 0.33,
        "Rinf": 6400,
        "R0": 270,
        "b": 8,
        "C[0]": 162400,
        "C[1]": 6750,
        "g[0]": 2800,
        "g[1]": 25,
        "m": 24,
        "K": 151
       }
hypothesis = "plane_stress"

material = mf.MFrontNonlinearMaterial("/home/vaishnav27/Fenics_Simulation/src/libBehaviour.so",
                                  "ChabocheViscoplasticity",
                                  hypothesis=hypothesis,
                                  material_properties=mat_prop)

problem = mf.MFrontNonlinearProblem(u, material, quadrature_degree=2, bcs=bc)
problem.initialize()
``` 

Next, I compute the residual for the dynamic problem and update the problem and compute the tangent form

```python
def m(u, u_):
    return rho*inner(u, u_)*problem.dx

def k(u_):
    return (sum([inner(g.variation(u_), f.function) for (f, g) in zip(problem.fluxes.values(), problem.gradients.values())]) * problem.dx)

def c(u, u_):
    return eta_m*m(u, u_) + eta_k*k(u_)

def Wext(u_):
    return loading*dot(n, u_)*ds(RIGHT_EDGE)

def update_a(u, u_old, v_old, a_old, ufl=True):
    if ufl:
        dt_ = dt
        beta_ = beta
    else:
        dt_ = float(dt)
        beta_ = float(beta)
    return (u-u_old-dt_*v_old)/beta_/dt_**2 - (1-2*beta_)/2/beta_*a_old

def update_v(a, u_old, v_old, a_old, ufl=True):
    if ufl:
        dt_ = dt
        gamma_ = gamma
    else:
        dt_ = float(dt)
        gamma_ = float(gamma)
    return v_old + dt_*((1-gamma_)*a_old + gamma_*a)

def update_fields(u, u_old, v_old, a_old):
    u_vec, u0_vec = u.vector(), u_old.vector()
    v0_vec, a0_vec = v_old.vector(), a_old.vector()

    a_vec = update_a(u_vec, u0_vec, v0_vec, a0_vec, ufl=False)
    v_vec = update_v(a_vec, u0_vec, v0_vec, a0_vec, ufl=False)

    v_old.vector()[:], a_old.vector()[:] = v_vec, a_vec
    u_old.vector()[:] = u.vector()

def avg(x_old, x_new, alpha):
    return alpha*x_old + (1-alpha)*x_new

a_new = update_a(du, u_old, v_old, a_old, ufl=True)
v_new = update_v(a_new, u_old, v_old, a_old, ufl=True)

residual = m(avg(a_old, a_new, alpha_m), u_) + c(avg(v_old, v_new, alpha_f), u_) + k(u_)
if not DISP_LOAD:
    residual -= Wext(u_)

problem.residual = residual
problem.compute_tangent_form()
``` 

But on the first increment I get the following error upon calling `problem.solve(u.vector())`

![image](https://github.com/user-attachments/assets/e8c579d9-0948-407d-a237-5ae60cb03b72)

I can integrate the material behaviour in the static case of the same problem setup without any issues. Could you please let me know if I'm missing something?

I am attaching the files [here](https://github.com/user-attachments/files/16785704/Code_Ref.zip)

**Comment by thelfer** [comment, marked as answer]
It seems that you pass the time increment as a  `Constant` when a floatting point number is expected.

**Reply by VaishnavBhaskaran** [reply]
Yes. That was causing the issue. Changing the time increment to float type resolved this error. Also the `TrialFunction` should be changed to `Function` here since it's a nonlinear problem.

Everything is running fine now. Thank you so much Dr. Helfer!

**Reply by thelfer** [reply]
Let me know if you have nice results, it's always good to illustrate MFront' usage.
Do not hesitate to participate to the MFront User Meeting next year (this year would be a bit ambitious and I already have too many speakers :))
Best,

**Reply by VaishnavBhaskaran** [reply]
Sure. I will definitely update about my work here.

I would gladly participate in the next year's meeting. Could you please share a tentative date?

**Reply by thelfer** [reply]
Not yet. I start organizing the meeting around May :) But if you send me an e-mail at `tfel-contact <at> cea.fr`, I'll add you to the mailling list.

**Reply by VaishnavBhaskaran** [reply]
Okay. I'll send you an email. Thank you!

**Reply by VaishnavBhaskaran** [reply]
Hello Dr. Helfer, I have sent an email to the above mail address. I also faced a strange behaviour with the time increments in the same problem setup. I am trying to perform an impact loading on a plate subjected to uniaxial tension. But the simulation does not converge for a time duration of `1e-1s` or less for the below load. Increasing the time increments did not change the non-convergence issue. It works if I increase the time to `1s` or more.

![image](https://github.com/user-attachments/assets/3fe7b32d-0eaa-4b3b-96be-b356f850185b)

But if I change the load as shown below, the simulation runs (for time > 1s) till the peak load and does not converge in the next time increment. I observed that the difference between the loads in these time increments are very small. I ideally wanted to apply a sinusoidal or a pulse load but couldn't do it because of this issue.

![image](https://github.com/user-attachments/assets/12cdf019-cce3-4234-9038-23846970b354)

I also played with the damping parameters, but it did not solve this issue. Could you please suggest a direction to resolve this?

**Reply by thelfer** [reply]
Can you tell me what did not converged ? The equilibrium algorithm or the MFront behaviour ?

**Reply by VaishnavBhaskaran** [reply]
The solver does not converge as shown below after reaching maximum iterations.

![image](https://github.com/user-attachments/assets/2ecc983d-fd5f-47f0-9912-e308ae3676f3)

**Reply by thelfer** [reply]
Can you show the evolution of the residual ?

**Reply by VaishnavBhaskaran** [reply]
![image](https://github.com/user-attachments/assets/0439af38-3245-4513-a49a-05e6b9776f49)

The values inside the square brackets are the loads.

**Reply by thelfer** [reply]
The stagnation of the residual is strange, in particular when the previous iteration seems very nice (could it be almost purely elastic ?)

**Reply by thelfer** [reply]
Would it be meaningful to make the same simulation without inertial terms ?

**Reply by VaishnavBhaskaran** [reply]
The previous iteration is purely elastic (not even close to the yield limit). 

![image](https://github.com/user-attachments/assets/1019dddb-c4c2-437e-9981-de1616f63a79)

I excluded the inertial and damping terms and ran it for `1e-2s`, but the solver does not converge when the load reached close to the peak, where the difference in load between successive increments is small as shown in the image below

![image](https://github.com/user-attachments/assets/45948ad0-7deb-45b3-a609-80973c60808d)

**Reply by thelfer** [reply]
So there something really bad regarding the consistent tangent operator here.
Could you attach the file (once again)

**Reply by VaishnavBhaskaran** [reply]
Sure. No problem! Please find the files attached [here](https://github.com/user-attachments/files/16995654/Dynamic_Chaboche.zip). The code contains the inertial terms. I have made some notes with line numbers at the end in the `Small_Strain_Plasticity_Dynamic.py` file for your reference.

**Reply by thelfer** [reply]
The criterion in   `ChabocheViscoplasticity.mfront` is much too high. You shoud use a value of `1e-12` or better `1e-14` as follows:

~~~c++
@Epsilon 1e-14;
~~~

**Reply by VaishnavBhaskaran** [reply]
Is the criterion `Epsilon` different from the `eps` defined in the later section of the code?

```python
@Integrator{
  const real two_third     = 2/real(3);
  constexpr const real eps = real(1.e-7);
``` 

I changed both of them just in case. So now I can simulate for `1e-2s` with the following load
![Plot_Load_Time](https://github.com/user-attachments/assets/21695457-b3d7-401f-9d1a-4f0dc16d59ad)
and get the output clearly as shown below
![Plot_Stress_Strain](https://github.com/user-attachments/assets/a530e670-3189-4fd2-93e7-5d5714368fdf)
I am able to get a similar output if I include the damping matrix alone. But if I include the mass matrix in the residual it doesn't converge in the second iteration itself. Is it because of the large acceleration (`mm/s^2`)?
Excluding mass matrix:
![image](https://github.com/user-attachments/assets/65cc4857-a386-4598-bb99-c4f310ebba6e)

Including mass matrix:
![image](https://github.com/user-attachments/assets/b4202d9f-f660-40ac-a960-02b7eb1b1a0d)

And could you please let me know how did you conclude that the value of `Epsilon` is causing this error? Just asking out of curiosity 😃

**Reply by thelfer** [reply]
You may want to use:

~~~c++
constexpr const real eps = real(1.e-12);
~~~

or lower.

A small value of the tolerance leads to:

1. numerically unstable results.
2. poor consistent tangent operator.

To understand the second point, you have to bear in mind that the consistent tangent operator is computed automatically using the implicit function theorem. But the assumptions of this theorem are only approximatively met.

**Reply by thelfer** [reply]
I let you investigate the influence of the dynamic effects.

**Reply by VaishnavBhaskaran** [reply]
Okay. Thank you Dr. Helfer! I will look into the dynamic effects' influence.

**Reply by VaishnavBhaskaran** [reply]
Hello Dr. Helfer,

I somehow couldn't find the reason behind the large accelerations. I am able to get a solution if I compute the residual as shown below excluding the mass matrix (was just trying to find the source of the problem)
```python
residual = self.c(self.avg(self.v_old, v_new, self.alpha_f), self.u_) + self.k(self.u_) - self.Wext(self.u_)
``` 
But if I compute the residual with the mass matrix, I get large accelerations in the first few iterations and the solver doesn't converge.
```python
residual = self.m(self.avg(self.a_old, a_new, self.alpha_m), self.u_)  + self.k(self.u_) - self.Wext(self.u_)
``` 
I have tried the following to identify the source of the issue:

- Simulated for longer time period (t>1s) instead of my impact time (1ms) with different types of load
- Changed the parameters of the integration scheme
- Changed the size of time increments

To check if my problem setup is correct, I modified the geometry and loading conditions in the Elastodynamics tutorial to meet my requirements and I am able to get the graph shown below, which is what I expected
![image](https://github.com/user-attachments/assets/3d18b18b-ebc9-4feb-bd53-aed076c54a7e)

I couldn't figure out if it's the MFront file or something else causing this. Could you please suggest me a solution?

**Reply by thelfer** [reply]
Did you try to set R0 and Rinf to a high value to see what happens with a purely elastic answer

**Reply by VaishnavBhaskaran** [reply]
Yes. I tried increasing them but still get the same issue.

These are the values I am using for my simulation. The material is Aluminium.
```python
RHO = 8000
MAT_PROP = {"YoungModulus": 185000,
        "PoissonRatio": 0.33,
        "Rinf": 6400,
        "R0": 270,
        "b": 8,
        "C[0]": 162400,
        "C[1]": 6750,
        "g[0]": 2800,
        "g[1]": 25,
        "m": 4.22,
        "K": 63.12
       }
```

**Comment by thelfer** [comment]
Did you try with:

```python
RHO = 8000
MAT_PROP = {"YoungModulus": 185000,
        "PoissonRatio": 0.33,
        "Rinf": 6400e9,
        "R0": 270e9,
        "b": 8,
        "C[0]": 162400,
        "C[1]": 6750,
        "g[0]": 2800,
        "g[1]": 25,
        "m": 4.22,
        "K": 63.12
       }
 ```

**Reply by VaishnavBhaskaran** [reply]
Yes. Irrespective of how high R0 and Rinf are, the stress value reaches close to the yield limit in a couple of iterations itself and doesn't converge. The accelerations scale based on this as well.

Below is the stress-strain plot just before the divergence, for the above mentioned parameters. It only runs for two iterations. I change the time increment, then it runs for one or two more iterations and then the issue arises again.
![image](https://github.com/user-attachments/assets/7cac3240-41ea-4be6-aea9-fda47485a43d)
Here is the log of the simulation. The underlined values are the accelerations. The list printed before that are the velocities.

![image](https://github.com/user-attachments/assets/97902b6c-0734-49c5-a2e7-69a080250961)

**Reply by thelfer** [reply]
With an elastic material, you shall be able to reproduce the Elastodynamics tutorial. I would investigate on that direction. I can't help more

**Reply by VaishnavBhaskaran** [reply]
Okay. I will work in this direction. Thank you Dr. Helfer! I will update here when I find a solution.
